hbase-server/src/test/java/org/apache/hadoop/hbase/client/AbstractTestUpdateConfiguration.java

```diff
@@ -0,0 +1,102 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.client;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.hadoop.hbase.HBaseTestingUtility;
+import org.apache.hadoop.hbase.util.JVMClusterUtil.RegionServerThread;
+
+/**
+ * Base class to test Configuration Update logic. It wraps up things needed to
+ * test configuration change and provides utility methods for test cluster setup,
+ * updating/restoring configuration file.
+ */
+public abstract class AbstractTestUpdateConfiguration {
+  private static final int SERVER_CONFIG_INDEX  = 0;
+  private static final int OVERWRITE_SERVER_CONFIG_INDEX  = 1;
+  private static final int BACKUP_SERVER_CONFIG_INDEX  = 2;
+  private static List<Path> newCnfPathes = new ArrayList<>();
+
+  protected static void setUpConfigurationFiles(final HBaseTestingUtility testUtil)
+    throws Exception {
+    // Before this change, the test will update hbase-site.xml under target/test-classes and
+    // trigger a config reload. Since target/test-classes/hbase-site.xml is being used by
+    // other testing cases at the same time, this update will break other testing cases so it will
+    // be flakey in nature.
+    // To avoid this, the change is to make target/test-classes/hbase-site.xml immutable. A new
+    // hbase-site.xml will be created under its test data directory, i.e,
+    // hbase-server/target/test-data/UUID, this new file will be added as a resource for the
+    // config, new update will be applied to this new file and only visible to this specific test
+    // case. The target/test-classes/hbase-site.xml will not be changed during the test.
+
+    String absoluteDataPath = testUtil.getDataTestDir().toString();
+
+    // Copy Files from target/test-classes to test-data directory, this is to avoid messing up
+    // the config file under target/test-class, which is supposed to be immutable.
+    Files.createDirectories(Paths.get(absoluteDataPath));
+
+    List<String> confs = Arrays.asList("/hbase-site.xml", "/hbase-site2.xml", "/hbase-site3.xml");
+    for (String conf : confs) {
+      Path cnfPath = Paths.get("target/test-classes" + conf);
```

Comment:
Again you're concatenating strings instead of using the API as provided. How about

```java
final Path confPath = Paths.get("target", "test-classes", "hbase-site.xml");
```

Reply:
Done.